xds/src/main/java/io/grpc/xds/ServerXdsClient.java

```diff
@@ -0,0 +1,240 @@
+        }
+      }
+    } catch (InvalidProtocolBufferException e) {
+      getLogger().log(XdsLogLevel.WARNING, "Failed to unpack Listeners in LDS response {0}", e);
+      nackResponse(ResourceType.LDS, nonce, "Malformed LDS response: " + e);
+      return;
+    }
+    ListenerUpdate listenerUpdate = null;
+    if (requestedListener != null) {
+      if (ldsRespTimer != null) {
+        ldsRespTimer.cancel();
+        ldsRespTimer = null;
+      }
+      try {
+        listenerUpdate = ListenerUpdate.newBuilder()
+            .setListener(EnvoyServerProtoData.Listener.fromEnvoyProtoListener(requestedListener))
+            .build();
+      } catch (InvalidProtocolBufferException e) {
+        getLogger().log(XdsLogLevel.WARNING, "Failed to unpack Listener in LDS response {0}", e);
+        nackResponse(ResourceType.LDS, nonce, "Malformed LDS response: " + e);
+        return;
+      }
+    } else {
+      if (ldsRespTimer == null) {
+        listenerWatcher.onResourceDoesNotExist(":" + listenerPort);
+      }
+    }
+    ackResponse(ResourceType.LDS, versionInfo, nonce);
+    if (listenerUpdate != null) {
+      listenerWatcher.onListenerChanged(listenerUpdate);
+    }
+  }
+
+  private boolean isRequestedListener(Listener listener) {
+    // TODO(sanjaypujare): check listener.getName() once we know what xDS server returns
+    return isAddressMatching(listener.getAddress())
+        && hasMatchingFilter(listener.getFilterChainsList());
+  }
+
+  private boolean isAddressMatching(Address address) {
+    // TODO(sanjaypujare): check IP address once we know xDS server will include it
+    return address.hasSocketAddress()
+        && (address.getSocketAddress().getPortValue() == listenerPort);
+  }
+
+  private boolean hasMatchingFilter(List<FilterChain> filterChainsList) {
+    // TODO(sanjaypujare): if myIp to be checked against filterChainMatch.getPrefixRangesList()
+    for (FilterChain filterChain : filterChainsList) {
+      FilterChainMatch filterChainMatch = filterChain.getFilterChainMatch();
+
+      if (listenerPort == filterChainMatch.getDestinationPort().getValue()) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  @Override
+  protected void handleStreamClosed(Status error) {
+    cleanUpResourceTimer();
+    if (listenerWatcher != null) {
+      listenerWatcher.onError(error);
+    }
+  }
+
+  @Override
+  protected void handleStreamRestarted() {
+    if (listenerWatcher != null) {
+      ldsRespTimer =
+          syncContext
+              .schedule(
+                  new ListenerResourceFetchTimeoutTask(":" + listenerPort),
+                  INITIAL_RESOURCE_FETCH_TIMEOUT_SEC, TimeUnit.SECONDS, getTimeService());
+    }
+  }
+
+  @Override
+  protected void handleShutdown() {
+    cleanUpResourceTimer();
+  }
+
+  @Override
+  protected void runWithSynchronized(Runnable runnable) {
+    syncContext.execute(runnable);
+  }
+
+  private void cleanUpResourceTimer() {
```

Comment:
`ClientXdsClient` has `cleanUpResourceTimers`. We can combine the 2 into 1 abstract method `cleanUpResourceTimers` in `AbstractXdsClient` which might also simplify the code at call sites.

Reply:
No, `cleanUpResourceTimers` should be concrete implantation's own logic. The abstract layer does not aware how concrete implementations subscribing to resources. This `cleanUpResourceTimer` method is just a helper method in the concrete implementation, it's used for both XdsClient shutdown and `handleStreamClosed` callback.